### PR TITLE
Lims 4796: Testplans_popup

### PR DESCRIPTION
**Pull request by dianaeldemerdash**
This pull request included three test cases related to each others 
1-Make sure that the test plans & units in the edit mode same as in the test plan popup (LIMS-4796)
2-Make sure when you update the test plan by add more one, this update should reflected on the test plans popup (LIMS-8256) 
3-Make sure when you update the test plan by replace it by another one this update should reflect on the test plans popup (LIMS-4802)

```
 nosetests3 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test044_testplans_popup --tc-file=config.ini --tc=site.password:admin 
2020-06-14 04:51:42.972 | INFO     | api_testing.apis.base_api:info:93 - Get authorized api session.
2020-06-14 04:51:42.972 | INFO     | api_testing.apis.base_api:info:93 - admin:admin
2020-06-14 04:51:43.781 | INFO     | api_testing.apis.base_api:info:93 - session ID : XxkrtjycKL1JYSRh-L3JZf6d4B5BvkjCTAW_yLQGIyw .....
Orders: Test plan pop up Approach: Make sure the test plans ...         
2020-06-14 04:51:44.011 | INFO     | ui_testing.testcases.base_test:info:134 - Test case : test044_testplans_popup
2020-06-14 04:52:15.085 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:524 - wait until page is loaded
2020-06-14 04:52:16.824 | INFO     | api_testing.apis.base_api:info:93 - set order configuration
2020-06-14 04:52:18.000 | INFO     | api_testing.apis.base_api:info:93 - status code: 1
2020-06-14 04:52:18.001 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/orders/auto/generatedId
2020-06-14 04:52:18.730 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:18.730 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/testPlans
2020-06-14 04:52:19.994 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:19.997 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/materialTypes
2020-06-14 04:52:20.468 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:20.468 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/testPlans/get/964
2020-06-14 04:52:20.844 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:20.844 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/contacts
2020-06-14 04:52:21.700 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:21.701 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/orders
2020-06-14 04:52:22.341 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:52:22.341 | INFO     | ui_testing.testcases.base_test:info:134 - open the order record in the edit mode
2020-06-14 04:52:24.798 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:524 - wait until page is loaded
2020-06-14 04:52:26.805 | INFO     | ui_testing.pages.order_page:get_suborder_data:335 - getting main order data
2020-06-14 04:52:26.992 | INFO     | ui_testing.pages.base_pages:info:277 -  getting suborders data
2020-06-14 04:52:27.759 | INFO     | ui_testing.testcases.base_test:info:134 - get test plan name 
2020-06-14 04:52:28.104 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-06-14 04:52:33.156 | INFO     | ui_testing.testcases.base_test:info:134 - get test plan & test unit name from the test plan popup
2020-06-14 04:52:33.156 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  Get test plans page.
2020-06-14 04:52:35.183 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:524 - wait until page is loaded
2020-06-14 04:52:49.142 | INFO     | ui_testing.pages.base_pages:sleep_medium:47 -  Medium sleep.
2020-06-14 04:52:59.963 | INFO     | ui_testing.testcases.base_test:info:134 - get test unit name
2020-06-14 04:52:59.963 | INFO     | ui_testing.testcases.base_test:info:134 - assert that the test plan in the editmode same as the test plan in the test plan pop up
2020-06-14 04:52:59.964 | INFO     | ui_testing.testcases.base_test:info:134 - assert that the test unint in the edit mode same as the test unit in the test unit pop up 
2020-06-14 04:53:00.037 | INFO     | ui_testing.testcases.base_test:info:134 - TearDown.        
ok

----------------------------------------------------------------------
Ran 1 test in 76.027s

OK

 nosetests3 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test045_testplans_popup_after_edit_by_add --tc-file=config.ini --tc=site.password:admin 
2020-06-14 04:53:44.818 | INFO     | api_testing.apis.base_api:info:93 - Get authorized api session.
2020-06-14 04:53:44.818 | INFO     | api_testing.apis.base_api:info:93 - admin:admin
2020-06-14 04:53:45.611 | INFO     | api_testing.apis.base_api:info:93 - session ID : q6nlDm5pymskzKgzWijdcDkKc5ALFCJMIwKyTObkCVk .....
Orders: Test plan pop up  Approach: Make sure In case you edit the test plans ...       
2020-06-14 04:53:45.860 | INFO     | ui_testing.testcases.base_test:info:134 - Test case : test045_testplans_popup_after_edit_by_add
2020-06-14 04:54:17.327 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:524 - wait until page is loaded
2020-06-14 04:54:18.673 | INFO     | api_testing.apis.base_api:info:93 - set order configuration
2020-06-14 04:54:19.898 | INFO     | api_testing.apis.base_api:info:93 - status code: 1
2020-06-14 04:54:19.898 | INFO     | ui_testing.testcases.base_test:info:134 - Get completed test plan to upade by it with raw material type
2020-06-14 04:54:19.898 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/testPlans
2020-06-14 04:54:21.505 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:21.508 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/orders/auto/generatedId
2020-06-14 04:54:21.918 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:21.920 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/testPlans
2020-06-14 04:54:23.037 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:23.038 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/materialTypes
2020-06-14 04:54:23.392 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:23.392 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/testPlans/get/1157
2020-06-14 04:54:23.783 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:23.783 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/contacts
2020-06-14 04:54:24.776 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:24.777 | INFO     | api_testing.apis.base_api:info:93 - GET : https://automation.1lims.com/api/orders
2020-06-14 04:54:25.460 | INFO     | api_testing.apis.base_api:info:93 - Status code: 1
2020-06-14 04:54:25.461 | INFO     | ui_testing.testcases.base_test:info:134 - open the order record in the edit mode
2020-06-14 04:54:28.192 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:524 - wait until page is loaded
2020-06-14 04:54:29.841 | INFO     | ui_testing.testcases.base_test:info:134 - go to update the test plan by adding the completed one
2020-06-14 04:54:30.932 | INFO     | ui_testing.pages.base_pages:info:277 -  set test plan : ['6c6be8fbe5'] for 1 time(s)
2020-06-14 04:54:38.869 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-06-14 04:54:43.870 | INFO     | ui_testing.pages.base_pages:info:277 -  set test unit : [] for 0 time(s)
2020-06-14 04:54:43.870 | INFO     | ui_testing.pages.base_pages:info:277 -  save the changes
2020-06-14 04:54:43.871 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-06-14 04:54:46.185 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-06-14 04:54:48.312 | INFO     | ui_testing.pages.order_page:get_suborder_data:335 - getting main order data
2020-06-14 04:54:48.481 | INFO     | ui_testing.pages.base_pages:info:277 -  getting suborders data
2020-06-14 04:54:49.413 | INFO     | ui_testing.testcases.base_test:info:134 - Get the test plan name that I added it in the edit mode
2020-06-14 04:54:49.788 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-06-14 04:54:54.832 | INFO     | ui_testing.testcases.base_test:info:134 - assert that the test plan I added in the test plan popup 
2020-06-14 04:54:54.913 | INFO     | ui_testing.testcases.base_test:info:134 - TearDown.        
ok

----------------------------------------------------------------------
Ran 1 test in 69.054s

OK

nosetests3 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test046_testplans_popup_after_edit_by_replace --tc-file=config.ini --tc=site.password:admin 
2020-06-14 05:06:50.458 | INFO     | api_testing.apis.base_api:info:93 - Get authorized api session.
2020-06-14 05:06:50.458 | INFO     | api_testing.apis.base_api:info:93 - admin:admin
2020-06-14 05:06:51.253 | INFO     | api_testing.apis.base_api:info:93 - session ID : OWm43t7wTToc3ATPBGFYdvZjVm-5DagTo73vTxi8co4 .....
Orders: Test plan: Test unit pop up Approach: In case I delete test plan, make sure it ... SKIP: https://modeso.atlassian.net/browse/LIMSA-127

----------------------------------------------------------------------
Ran 1 test in 0.001s

OK (SKIP=1)

```

### CODE REVIEW

File: ui_testing/pages/order_page.py [hunk 1]
```diff
@@ -578,6 +578,25 @@ def get_sub_order_data_first_row(self, index=0):
         suborders = self.base_selenium.get_table_rows(element='order:suborder_table')
         suborder_row = suborders[index]
         suborder_data = self.get_suborder_data()
-        return  suborder_data
+        return suborder_data
+
+    def get_testplan_pop_up(self, sub_order_index=0, one_testplan=False, multiple_testplan=False):
```
Comment (0xIslamTaha) [inline ui_testing/pages/order_page.py:583]:
```suggestion
    def get_testplan_pop_up():
```

File: ui_testing/pages/order_page.py [hunk 2]
```diff
@@ -578,6 +578,25 @@ def get_sub_order_data_first_row(self, index=0):
...
+        self.base_selenium.click(element='order:testplan_popup_btn')
+        self.sleep_small()
+        if one_testplan == True:
+           element = self.base_selenium.find_element_by_xpath(xpath='//div[@class="col-md-12 border-bottom"]')
+           testplans_testunits_names = element.text
+           return testplans_testunits_names
+        if multiple_testplan == True:
+           element = self.base_selenium.find_element(element='order:popup_testplan')
+           testplans_testunits_names = element.text
+           return testplans_testunits_names
```
Comment (0xIslamTaha) [inline ui_testing/pages/order_page.py:597]:
```suggestion
        results = []
        elements = self.base_selenium.find_elements('order:popup-data') # class_name= 'border-bottom'
        for element in elements:
            test_plan, test_units = element.text.split('\n')[0], element.text.split('\n')[1:]
            results.append({'test_plan': test_plan, 'test_units': test_units})
        return results

```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 3]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        order, payload = self.orders_api.create_new_order()
+        self.info('open the order record in the edit mode')
+        self.orders_page.get_order_edit_page_by_id(id=order['order']['mainOrderId'])
+        testplan_name = self.order_page.get_sub_order_data_first_row()['suborders'][0]['testplans'][0]
+        self.info("get test plan name ".format(testplan_name))
+        testplans_testunits_names_in_popup = self.order_page.get_testplan_pop_up(one_testplan=True)
+        self.info("get test plan & test unit name from the test plan popup".format(testplans_testunits_names_in_popup))
+        self.testplan_page.get_test_plans_page()
+        self.testplan_page.search(testplan_name)
+        testunit_name = self.testplan_page.get_child_table_data()[0]['Test Unit Name']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2680]: use API

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 4]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        self.info("get test unit name".format(testunit_name))
+        self.info("assert that the test plan in the editmode same as the test plan in the test plan pop up".format(testplan_name, testplans_testunits_names_in_popup))
+        self.assertIn(testplan_name, testplans_testunits_names_in_popup)
+        self.info("assert that the test unint in the edit mode same as the test unit in the test unit pop up ".format(testunit_name, testplans_testunits_names_in_popup))
+        self.assertIn(testunit_name, testplans_testunits_names_in_popup)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2685]: update it according to the suggested `get_testplan_pop_up`

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 5]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        self.info('Get completed test plan to upade by it with raw material type')
+        testplan = \
+            self.test_plan_api.get_completed_testplans_with_material_and_same_article(material_type='Raw Material', article='', articleNo='')[0]
+        order, payload = self.orders_api.create_new_order(materialTypeId=1)
+        self.info('open the order record in the edit mode')
+        self.orders_page.get_order_edit_page_by_id(id=order['order']['mainOrderId'])
+        self.info('go to update the test plan by adding the completed one')
+        self.order_page.update_suborder(sub_order_index=0, test_plans=[testplan['testPlanName']])
+        self.order_page.save(save_btn='order:save')
+        testplan_name = self.order_page.get_sub_order_data_first_row()['suborders'][0]['testplans'][1]
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2702]:
```suggestion
        testplan_name = testplan['testPlanName']
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 6]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        self.info('Get completed test plan to upade by it with raw material type')
+        testplan = \
+            self.test_plan_api.get_completed_testplans_with_material_and_same_article(material_type='Raw Material',article='', articleNo='')[0]
+        order, payload = self.orders_api.create_new_order(materialTypeId=1)
+        self.info('open the order record in the edit mode')
+        self.orders_page.get_order_edit_page_by_id(id=order['order']['mainOrderId'])
+        self.info('go to update the test plan by adding the completed one')
+        self.order_page.update_suborder(sub_order_index=0, remove_old=True, test_plans=[testplan['testPlanName']])
+        self.order_page.save(save_btn='order:save')
+        testplan_name = self.order_page.get_sub_order_data_first_row()['suborders'][0]['testplans'][0]
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2724]:
```suggestion
        testplan_name = testplan['testPlanName']
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 7]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        self.info("Get the test plan name that I added it in the edit mode".format(testplan_name))
+        testplans_testunits_names_in_popup = self.order_page.get_testplan_pop_up(one_testplan=True)
+        self.info("assert that the test plan I added in the test plan popup ".format(testplan_name,
+                                                                                       testplans_testunits_names_in_popup))
+        self.assertIn(testplan_name, testplans_testunits_names_in_popup)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2729]:
```suggestion
        self.assertEqual(len(testplans_testunits_names_in_popup), 1)
        self.assertEqual(testplan_name, testplans_testunits_names_in_popup[0]['test_plan'])
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 8]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+        self.info('open the order record in the edit mode')
+        self.orders_page.get_order_edit_page_by_id(id=order['order']['mainOrderId'])
+        self.info('go to update the test plan by adding the completed one')
+        self.order_page.update_suborder(sub_order_index=0, test_plans=[testplan['testPlanName']])
+        self.order_page.save(save_btn='order:save')
+        testplan_name = self.order_page.get_sub_order_data_first_row()['suborders'][0]['testplans'][1]
+        self.info("Get the test plan name that I added it in the edit mode".format(testplan_name))
+        testplans_testunits_names_in_popup = self.order_page.get_testplan_pop_up(multiple_testplan=True)
+        self.info("assert that the test plan I added in the test plan popup ".format(testplan_name, testplans_testunits_names_in_popup))
+        self.assertIn(testplan_name, testplans_testunits_names_in_popup)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2706]: update it according to the new implementation of get_testplan_pop_up

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 9]
```diff
@@ -2659,3 +2661,71 @@ def test044_table_with_add_edit_single_row(self):
...
+    def test044_testplans_popup(self):
+        """
+        Orders: Test plan pop up Approach: Make sure the test plans
+        & units displayed on the test plans & units fields same as in the test plan pop up
+        LIMS-4796
+        """
+        order, payload = self.orders_api.create_new_order()
+        self.info('open the order record in the edit mode')
+        self.orders_page.get_order_edit_page_by_id(id=order['order']['mainOrderId'])
+        testplan_name = self.order_page.get_sub_order_data_first_row()['suborders'][0]['testplans'][0]
```
[the remark this answers is not in the document]
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2674]: Done

[the remark this answers is not in the document]
Comment (0xIslamTaha) [reply ui_testing/testcases/basic_tests/test06_orders.py:2674]: @dianaeldemerdash Did u try my suggestions?

File: ui_testing/pages/order_page.py [hunk 10]
```diff
@@ -578,6 +578,25 @@ def get_sub_order_data_first_row(self, index=0):
...
         suborder_row = suborders[index]
         suborder_data = self.get_suborder_data()
-        return  suborder_data
+        return suborder_data
+
+    def get_testplan_pop_up(self, sub_order_index=0, one_testplan=False, multiple_testplan=False):
+        suborder_table_rows = self.base_selenium.get_table_rows(
+            element='order:suborder_table')
+        suborder_row = suborder_table_rows[sub_order_index]
+        suborder_row.click()
```
[the remark this answers is not in the document]
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:587]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2674]: @0xIslamTaha yes I tried it from the get sub order data but I notice now that you want the test plan data from the create order using api

[the remark this answers is not in the document]
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2674]: Done

File: ui_testing/pages/order_page.py (hunk 2 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:597]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 4 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2685]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 3 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2680]: Done

File: ui_testing/pages/order_page.py (hunk 1 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:583]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 5 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2702]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 6 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2724]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 7 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2729]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 8 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py]: Done

**Pull request merged**